[JENKINS-60266, JEP-223] Add Jenkins.MANAGE permission and allow some item of /manage with it

aHenryJard opened the pull request:
See [JENKINS-60266](https://issues.jenkins-ci.org/browse/JENKINS-60266).
But also the JEP-223: https://github.com/jenkinsci/jep/tree/master/jep/223
And also the announcement in dev list: https://groups.google.com/forum/#!topic/jenkinsci-dev/lThSifB8G4k

[current WIP] This is an experimental feature, the new Jenkins.MANAGE permission is disabled by default (is equals to Jenkins.ADMINISTER).
This can be enabled with a system property "jenkins.permission.manage.enabled", for example in command line by adding `-Djenkins.permission.manage.enabled=true`.

### Proposed changelog entries

* Entry 1: Addition of a new permission Jenkins.MANAGE that allow a user to configure Jenkins without the need to be ADMINISTER. Move some management item to be allowed to this new MANAGE permission. This is an experimental feature, disabled by default that can be enabled with "jenkins.permission.manage.enabled" system property.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

Commit:
Add Jenkins.CONFIGURE permission and allow some item of /manage with this permission.

Commit:
Add Jenkins.CONFIGURE permission and allow some item of /manage with this permission.

Commit:
Annotation with Jira number for tests related to CONFIGURE.

Commit:
Annotation with Jira number for tests related to CONFIGURE.

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -5267,7 +5273,15 @@ private static void computeVersion(ServletContext context) {
     public static final Permission ADMINISTER = Permission.HUDSON_ADMINISTER;
+    /**
+     * Allows non-privilege escalating configuration permission for a Jenkins instance.  Actions which could result
+     * in a privilege  escalation (such as RUN_SCRIPTS) require explicit ADMINISTER permission
```
If `RUN_SCRIPTS` is now deprecated, why are you referring to it?

mikecirioli replied:
the deprecation is being handled in a separate PR, hence we act as if it is still around here...

aHenryJard replied:
For the record, the PR on RUN_SCRIPT is there https://github.com/jenkinsci/jenkins/pull/4365

jglick commented on core/src/main/java/jenkins/model/RunIdMigrator.java:
```diff
@@ -423,7 +423,7 @@ public String getUrlName() {
         @Override
         public Object getTarget() {
-            Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+            Jenkins.get().checkPermission(Jenkins.CONFIGURE);
```
Better revert.

aHenryJard replied:
reverted to ADMINISTER.

jglick commented on core/src/main/java/jenkins/security/ResourceDomainConfiguration.java:
```diff
@@ -82,7 +82,7 @@ public ResourceDomainConfiguration() {
     @POST
     public FormValidation doCheckUrl(@QueryParameter("url") String resourceRootUrlString) {
-        Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+        Jenkins.get().checkPermission(Jenkins.CONFIGURE);
```
Revert, this seems like an admin thing.

aHenryJard replied:
This method is used when the page "/configure" is opened, so if we want that some part of "Configure System" can be allowed to the CONFIGURE permission we have to allow it or move the call. I am looking which part of the "Configure system" page need that call.

jglick replied:
This ought to be moved to `/configureSecurity/` I think (change the category on the `GlobalConfiguration`).

aHenryJard replied:
changed to ADMINISTER, I still put in the todolist to look to move that check because the "/configure" page is not in error state anymore, but there is still some error on the UI and in the console. So there is still something to work on there.

jglick commented on core/src/main/resources/hudson/model/Computer/sidepanel.jelly:
```diff
@@ -37,7 +37,7 @@ THE SOFTWARE.
       <l:task href="${rootURL}/${it.url}load-statistics" icon="icon-monitor icon-md" title="${%Load Statistics}"/>
       <j:if test="${it.channel!=null}">
-        <l:task href="${rootURL}/${it.url}script" icon="icon-terminal icon-md" permission="${app.RUN_SCRIPTS}" title="${%Script Console}"/>
+        <l:task href="${rootURL}/${it.url}script" icon="icon-terminal icon-md" permission="${app.CONFIGURE}" title="${%Script Console}"/>
```
Uh, no! `ADMINISTER` for sure.

jglick reviewed (commented):
A quick pass over, not a real review.

EstherAF commented on core/src/main/java/hudson/logging/LogRecorderManager.java:
```diff
@@ -206,7 +206,7 @@ public static void init(Jenkins h) throws IOException {
     public Object getTarget() {
         if (!SKIP_PERMISSION_CHECK) {
-            Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+            Jenkins.get().checkPermission(Jenkins.CONFIGURE);
```
Probably a lot of plugins are logging secrets, especially at `FINE` level. 

I'm in the fence here, so I'm really happy to know your thoughts about this.

On one hand, logs can be sent to 3rd party systems, like datadog, etc, which would expose them. Given that, a plugin logging secrets can be considered a bug (security issue) in the actual plugin logging them. 

On the other hand, probably a lot of plugins are logging secrets, despite being a bug, so IMO, being realistic, ADMINISTER is the most secure option.

jglick replied:
I am also on the fence here.

mikecirioli replied:
After some internal discussion, the decision has been made to revert this back to ADMINISTER - the reasoning is that FINE level logging often contains all sorts of unexpected things, including sensitive values that we don't want to inadvertently expose.  As part of a future enchancement, we can consider creating a new logger that only exposes INFO level logging, and which could be viewable by someone with the CONFIGURE permission.

mikecirioli replied:
reverted

EstherAF commented on core/src/main/java/jenkins/management/SystemInfoLink.java:
```diff
@@ -52,4 +56,9 @@ public String getDescription() {
+
+    @Override
+    public Permission getRequiredPermission() {
+        return Jenkins.CONFIGURE;
```
I have the same concern but IMO credentials in env vars or system props can be considered a security design flaw. A pipeline can already access them, so if a user is storing credentials there, they are already exposed...

jglick replied:
> A pipeline can already access them

No it cannot, not without building on a master executor. See also: https://github.com/jenkinsci/script-security-plugin/blob/7bcf727f58e5f9de89424741ffe562245f1aa9a8/src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/blacklist#L64-L69

EstherAF commented:
Thanks @jglick, you did spot a lot of errors! :smile:

Commit:
Fix the missing permission on old data management for CONFIGURE

Commit:
Fix the missing permission on old data management for CONFIGURE

Commit:
Some changes regarding feedback on CONFIGURE.

Commit:
Some changes regarding feedback on CONFIGURE.

jtnord commented on core/src/main/java/hudson/model/Fingerprint.java:
```diff
@@ -996,7 +996,7 @@ public RangeItem(String name, RangeSet ranges) {
         for (Entry<String, RangeSet> e : usages.entrySet()) {
             final String itemName = e.getKey();
-            if (instance.hasPermission(Jenkins.ADMINISTER) || canDiscoverItem(itemName)) {
+            if (instance.hasPermission(Jenkins.CONFIGURE) || canDiscoverItem(itemName)) {
```
> Does Jenkins.CONFIGURE imply full access to all Items?

No.

jglick replied:
So…this looks like a minor vulnerability then?

jtnord replied:
> So…this looks like a minor vulnerability then?

Yes, this is something that needs to be reverted.

// @aHenryJard

aHenryJard replied:
reverted.

Commit:
Some more changes between ADMINISTER and CONFIGURE.

Commit:
Some more changes between ADMINISTER and CONFIGURE.

jtnord commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -1911,7 +1911,9 @@ public View getPrimaryView() {
 
     public void setPrimaryView(@Nonnull View v) {
-        this.primaryView = v.getViewName();
+        if (hasPermission(ADMINISTER)) {
```
> Why is this not CONFIGURE?

You would not be able to create Views (or even see them without other Permissions) so although this appears strange, I think it is technically correct.  However yes - this should `checkPermission(ADMINISTER)` so that the correct Exception is thrown.

aHenryJard replied:
changed to checkPermission.

jtnord replied:
Looking at the calling code I think this is not quite correct.
`GlobalDefaultViewConfiguration` should not try setting this if it doesn't have the permissions as this will always be sent in a save() IIUC causing saves of global configuration to fail?

Perhaps `GlobalConfiguration` needs to be enhanced to have `GlobalSensitiveConfiguration` or the like that a configure user will not be able to adjust?

aHenryJard committed:
Merge pull request #1 from aHenryJard/fix-old-data-missing-perm

Fix the missing permission on old data management for CONFIGURE

aHenryJard committed:
Some changes regarding feedback this PR for CONFIGURE.

Back to ADMINISTER for some sensible methods.

aHenryJard commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -1411,7 +1411,7 @@ public Api getApi() {
     public void doDumpExportTable( StaplerRequest req, StaplerResponse rsp ) throws IOException, ServletException, InterruptedException {
         // this is a debug probe and may expose sensitive information
-        checkPermission(Jenkins.ADMINISTER);
+        checkPermission(Jenkins.CONFIGURE);
```
ok, it's done.

aHenryJard commented on core/src/main/java/hudson/model/ManagementLink.java:
```diff
@@ -115,7 +115,7 @@ public boolean getRequiresConfirmation() {
      * @return permission required for user to access this management link, in addition to {@link Jenkins#ADMINISTER}
```
Modified.

aHenryJard commented on core/src/main/java/hudson/views/GlobalDefaultViewConfiguration.java:
```diff
@@ -51,7 +51,9 @@ public boolean configure(StaplerRequest req, JSONObject json) throws FormExcepti
         } else {
             // Fallback if the view is not specified
-            j.setPrimaryView(j.getViews().iterator().next());
+            if(j.getViews().iterator().hasNext()) {
```
I had an error page using a user with CONFIGURE permission because this was null and fires a NullPointerException. I don't remember which page exactly, I can test again. But maybe the issue is that the user with CONFIGURE should have a primaryView set instead ? I add this to the todo list.

jglick replied:
Related to #4348?

jvz replied:
You should store the Iterator in a variable. Technically, checking hasNext() and next() on two separate iterators doesn't make sense.

mikecirioli committed:
[JENKINS-60266] revert back to require Jenkins.ADMINISTER when viewing
logs

mikecirioli committed:
[JENKINS-60266] revert back to require Jenkins.ADMINISTER when viewing
logs

mikecirioli committed:
Add a filter for GlobalConfiguration items that cleanly hides those
that are not accessible to a user without sufficient permission

mikecirioli committed:
Add a filter for GlobalConfiguration items that cleanly hides those
that are not accessible to a user without sufficient permission

mikecirioli committed:
Move permission filter to a more logical location

mikecirioli committed:
Move permission filter to a more logical location

mikecirioli committed:
nit: fixup javadoc

mikecirioli committed:
nit: fixup javadoc

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
Overide to allow Jenkins.CONFIGURE

mikecirioli committed:
fix faulty logic in filter permission check

mikecirioli committed:
fix faulty logic in filter permission check

mikecirioli committed:
address code style nits

mikecirioli committed:
address code style nits

Commit:
[JENKINS-60266] Fix test about the permissions required to get the master's envvars

Commit:
[JENKINS-60266] Fix test about the permissions required to get the master's envvars

Commit:
[JENKINS-60266] Allow modifying Default View with CONFIGURE permission

Commit:
[JENKINS-60266] Allow modifying Default View with CONFIGURE permission

Commit:
[JENKINS-60266] Don't allow to edit Resource Root URL or Jenkins URL with CONFIGURE permission

Commit:
[JENKINS-60266] Don't allow to edit Resource Root URL or Jenkins URL with CONFIGURE permission

Commit:
[JENKINS-60266] Adress feedback: Revert some CONFIGURE permissions

Commit:
[JENKINS-60266] Adress feedback: Revert some CONFIGURE permissions

Commit:
[JENKINS-60266] Revert allowing CONFIGURE to access the logs

Commit:
[JENKINS-60266] Revert allowing CONFIGURE to access the logs

mdb23inc commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -1411,7 +1411,7 @@ public Api getApi() {
      */
     public void doDumpExportTable( StaplerRequest req, StaplerResponse rsp ) throws IOException, ServletException, InterruptedException {
         // this is a debug probe and may expose sensitive information
-        checkPermission(Jenkins.ADMINISTER);
```
ublic void doDumpExportTable( StaplerRequest req, StaplerResponse rsp ) throws IOException, ServletException, InterruptedException {
        // this is a debug probe and may expose sensitive information
        checkPermission(Jenkins.ADMINISTER);

mikecirioli committed:
Fix failing test (#4)

Fix failing test

oleg-nenashev commented:
I plan to finally start reviewing it after Jan 10 or so

mikecirioli committed:
CONFIGURE needs overal read (fix failing test)

mikecirioli committed:
CONFIGURE needs overal read (fix failing test)

mikecirioli committed:
Add a filter for GlobalConfiguration items based on permission (#3)

Add a filter for GlobalConfiguration items based on permission

jvz commented on core/src/main/java/jenkins/model/MasterBuildConfiguration.java:
```diff
@@ -51,12 +51,15 @@ public boolean configure(StaplerRequest req, JSONObject json) throws FormExcepti
-                throw new FormException(Messages.Hudson_Computer_IncorrectNumberOfExecutors(),"numExecutors");
+            if(j.hasPermission(Jenkins.ADMINISTER)){
+                String num = json.getString("numExecutors");
+                if (!num.matches("\\d+")) {
```
Nit: I know you're just refactoring here, but this could be done a bit more easily via `Integer.parseInt()` and catching the exception.

mikecirioli committed:
Rename permission Jenkins.CONFIGURE to Jenkins.MANAGE

mikecirioli committed:
Rename permission Jenkins.CONFIGURE to Jenkins.MANAGE

aHenryJard committed:
Merge pull request #5 from aHenryJard/rename-permission

Rename Jenkins.CONFIGURE permission to Jenkins.MANAGE

mikecirioli committed:
rename Jenkins.CONFIGURE permission to Jenkins,MANAGE in jelly files

mikecirioli committed:
rename Jenkins.CONFIGURE permission to Jenkins,MANAGE in jelly files

mikecirioli committed:
rename Jenkins.CONFIGURE permission to Jenkins.MANAGE in jelly fil… (#6)

rename Jenkins.CONFIGURE permission to Jenkins.MANAGE in jelly files

mikecirioli committed:
Do not show the "Plugin Dependencies" section if a user does not have
Jenkins.ADMINISTER permission

mikecirioli committed:
Do not show the "Plugin Dependencies" section if a user does not have
Jenkins.ADMINISTER permission

mikecirioli committed:
fixup broken tests so they look for the correct permission name

mikecirioli committed:
fixup broken tests so they look for the correct permission name

mikecirioli committed:
allow Jenkins.MANAGE users to view plugin depenencies in AboutJenkins
page

mikecirioli committed:
allow Jenkins.MANAGE users to view plugin depenencies in AboutJenkins
page

mikecirioli committed:
Don't dump stacktrace if user does not have the required permission

mikecirioli committed:
Don't dump stacktrace if user does not have the required permission

mikecirioli committed:
Allow user with Jenkins.MANAGE permission to view plugin dependenc… (#7)

Allow user with Jenkins.MANAGE permission to view plugin dependencies from the "About Jenkins" page

mikecirioli committed:
remove unneeded code

mikecirioli committed:
remove unneeded code

mikecirioli committed:
Revert "remove unneeded code"

This reverts commit b52803771ff8024830764ba724a03453e91c2c92.

mikecirioli committed:
Revert "remove unneeded code"

This reverts commit b52803771ff8024830764ba724a03453e91c2c92.

mikecirioli committed:
nit

mikecirioli committed:
nit

mikecirioli commented:
note: current failing test appears to be a flake

mikecirioli committed:
Merge branch 'configure-permission' of github.com:aHenryJard/jenkins into configure-permission
rebase and pull in current jenkins changes

timja reviewed (commented):
Looks like something went wrong in a merge, this PR has unrelated changes in it

Commit:
Revert "Merge branch 'configure-permission' of github.com:aHenryJard/jenkins into configure-permission"

This reverts commit 6ea42a0dce68575c20c7300d2d1a16c37201430a, reversing
changes made to 2a5989ec98f12ad5df95c50f86ef71dedf373af6.

aHenryJard commented on test/src/test/java/hudson/model/labels/LabelAtomPropertyTest.java:
```diff
@@ -68,4 +75,55 @@ public void configRoundtrip() throws Exception {
+    @Issue("JENKINS-60266")
+    @Test
+    public void configAllowedWithConfigurePermission() throws Exception {
+        final String CONFIGURATOR = "configurator";
```
@mikecirioli @EstherAF I think we can rename that user to "Manager" os something similar, what do you think ?

aHenryJard commented:
@jglick @daniel-beck @oleg-nenashev @batmat This PR commit history began to be quite messy, what is the usage to get a cleaner history ? Should we rewrite the history, maybe squash all to one commit or create a new PR because this one was a draft. What do you think ?

cc @EstherAF @mikecirioli

timja commented:
We can squash on merge it’s fine

Commit:
Some CONFIGURE left, mainly in test comment.

Commit:
Text error in comment

aHenryJard committed:
Merge pull request #8 from aHenryJard/config-perm-pr

Some CONFIGURE left, mainly in test comment and addressing PR feedback on setPrimaryView.

aHenryJard committed:
Adding a system property to activate MANAGE permission

Co-authored-by: Esther Álvarez Feijoo <5547479+EstherAF@users.noreply.github.com>

Commit:
Fix some test after adding FT flag to disable MANAGE

Commit:
rename some tests about MANAGE permission

aHenryJard committed:
Revert unrelated changes to test files.

aHenryJard committed:
Fix JenkinsTest and rename other test with CONFIGURE into MANAGE

Commit:
Apply merge from master.

aHenryJard committed:
Apply merge from master and resolve conflicts.

Commit:
Revert "Apply merge from master and resolve conflicts."

This reverts commit 4122f5d1798cae9d2def3f95ff68db0b5c21ba7e, reversing
changes made to 312955a1b079e6cdab1a70a5a8df3e9614665598.

EstherAF commented:
FTR, the git history of this PR has been messed up at some point, so I've created a new PR in order to avoid rewriting the history in this one. 

#4483

We may close this and follow in the other one, which would be more easy to review

daniel-beck commented:
Superseded by #4483.
